_rules/html-page-lang-b5c3f8.md

```diff
@@ -19,30 +19,25 @@ authors:
 
 ## Applicability
 
-The root element of the [web page](https://www.w3.org/TR/WCAG21/#dfn-web-page-s), if it is an `html` element.
-
-**Note**: Documents embedded into other documents, such as through `iframe` or `object` elements are not applicable because they are not web pages according to the definition in WCAG.
+This rule applies to any [document element](https://dom.spec.whatwg.org/#document-element) if it is an `html` element that has a [node document](https://dom.spec.whatwg.org/#concept-node-document) with a [content type](https://dom.spec.whatwg.org/#concept-document-content-type) of `text/html`.
```

Comment:
This rule shouldn't be applicable to document elements that are embedded into frames / objects. That's the reason for the original wording. This problem exists with the other rules too.

Reply:
This change was entirely intentional, see https://github.com/act-rules/act-rules.github.io/pull/934#discussion_r327576703.

Reply:
To illustrate the issue, consider this example:

```html
<!doctype html>
<html lang=da>
  <p>Hej med dig</p>
  <iframe srcdoc="<!doctype html><p>Hej med dig</p>"></iframe>
</html>
```

VoiceOver correctly announces the outer "Hej med dig" in Danish, but the inner "Hej med dig" is announced in my default system locale, English. The following works as expected:

```html
<!doctype html>
<html lang=da>
  <p>Hej med dig</p>
  <iframe srcdoc="<!doctype html><html lang=da><p>Hej med dig</p>"></iframe>
</html>
```

The rule should therefore absolutely be applicable to nested browsing contexts as well.

Reply:
I see your point. Even so, the success criterion this tests is whether or not the language of the page is set, and a web page is defined as a non-embedded resource. What you're pointing out could be considered a failure of 3.1.2, but I don't think that's a failure of 3.1.1.

Reply:
I think there's a bug in WCAG here with language of page / language of parts - because it uses a different definition for 'page' than HTML. The HTML language computation does not use WCAG's definition of web page when computing language. 

There's similar problem with `longdesc` where accessing the URL referenced by longdesc in Firefox is treated as page navigation to a new page, but WCAG defines long descriptions as part of the full web page:
https://www.w3.org/TR/UNDERSTANDING-WCAG20/conformance.html#uc-conf-req2-head

To see this in practice:
- open https://www.davidmacd.com/test/longdesc.html in Firefox
- right click on the first image
 - choose the 'View Description' command  

Moving this in a different PR makes sense since this is going to be hard to resolve.

Reply:
I don't think the definition of "web page" is the issue here; I think the definition is perfectly fine. Our individual readings of the WCAG differ though and that's really were I think the issue lies. Let me try to visualise it. First, here's hopefully something we can all agree on:

```
+---------------------+
|Web page             |
|                     |
|         +----------+|
|         |Embed     ||
|         |          ||
|         +----------+|
+---------------------+
```

The above shows a web page with an embedded resource. From the perspective of WCAG, the embedded resource forms a subset of the web page, but is not a web page in its own right; only the superset is. My hope is that you're nodding in agreement to this statement.

So, let's pretend for a second that the web page does not embed a resource and we've set its language to `da`:

```
+---------------------+
|Web page             |
|lang: da             |
|                     |
|                     |
|                     |
|                     |
+---------------------+
```

Assuming that the web page is indeed authored in Danish, this passes [success criterion 3.1.1](https://www.w3.org/TR/WCAG/#language-of-page) as the _default_ human language of the web page can be programmatically determined. 

Keeping the above in mind, consider this: What if, for whatever reason, the default human language only applied to _half_ the web page and we needed to do _something_ additional to also have it apply to the other half? I hope we can agree that without doing that additional _something_, the web page does not pass success criterion 3.1.1.

Now, let's take a long, hard look at this:

```
+---------------------+
|Web page             |
|lang: da             |
|         +----------+|
|         |Embed     ||
|         |lang: ?   ||
|         +----------+|
+---------------------+
```

This is really just a graphical depiction of the code I posted in https://github.com/act-rules/act-rules.github.io/pull/934#discussion_r328210245. Here, the default human language can be determined for everything but the embedded resource. For the sake of argument, let's pretend that the embedded resource contains half the content of the entire web page. This is _precisely_ the situation I described previously; the default human language can only be determined for _half_ the web page. Hopefully, we've all already agreed that this does _not_ pass success criterion 3.1.1.

Reply:
@dd8; I disagree this is a "bug". The choice for the definition of web page is very much intentional. We can discuss whether or not they should have chose a different term, such as "non-embedded web page", but that's beside the point that this SC was designed to only apply to the top-level document.

@kasperisager; I see where you're coming from, but I don't share your interpretation. The way I see it is this: If a Danish page has an English ad embedded in a div, that's a failure of 3.1.2, correct? Now take that ad and put it into an iframe with `html[lang=da]`. Still a failure of 3.1.2, and not of 3.1.1, right? So if that's the case, why then, if we were to remove that incorrect lang=da and not replace it, would it suddenly become a failure of 3.1.1? That is still the same piece of content, where the language of a section of the page can't be programatically determined. 

Regardless of what we conclude on, this change shouldn't be in a PR about xml:lang. We should not have one PR making multiple changes to the rule. Can you please undo it, and either open an issue, or create a different PR.

Reply:
> If a Danish page has an English ad embedded in a div, that's a failure of 3.1.2, correct?

Yes, since the _default language_ of that ad would be Danish.

> Now take that ad and put it into an iframe with html[lang=da]. Still a failure of 3.1.2, and not of 3.1.1, right?

Yes, because the _default language_ is _still_ Danish.

> So if that's the case, why then, if we were to remove that incorrect lang=da and not replace it, would it suddenly become a failure of 3.1.1?

Because now the _default language_ of the ad can no longer be determined programmatically and the web page therefore fails 3.1.1.

Regardless, I'll put this change into a separate PR and adjust the applicability in this PR.

Reply:
@DagfinnRomen and @ShadowBB I'd love your take on this as well.

Reply:
@kasperisager Are we moving this discussion to a separate PR?

According to WCAG, a web page is a non-embedded resource (...you know the rest, I won't paste the definition here) including all embedded images and media. The way I see it, a web page with an embedded iframe, would include the iframe as a part of the page. However, the content in this iframe would itself not count as a separate webpage.

I also agree with Kasper in https://github.com/act-rules/act-rules.github.io/pull/934#discussion_r329500292

Reply:
@kasperisager I agree with Wilco his interpretation. I would state that indeed the default language of the ad would be Danish. But 3.1.1 doesn’t requires that the default language of the ad has to be programmatically determinable. It only requires that the default language of the page can be programmatically determined.

Reply:
> It only requires that the default language of the page can be programmatically determined.

@ShadowBB That's exactly my point in https://github.com/act-rules/act-rules.github.io/pull/934#discussion_r329500292; the default language of the page _cannot_ be programmatically determined. Part of it can, yes, but not the part contained within the ad. It's really no different than talking about conformance of a page; if all non-embedded content on a page conforms to the WCAG, but some or all embedded content does not, the page as a whole does not conform to the WCAG.

In general, I'm having a really hard time understanding how one can regard a glass as full if only the top half is filled, but the bottom half isn't. Disregarding the physical impossibility of that analogy, I don't see it as any different.